subsys/bluetooth/audio/ascs.c

```diff
@@ -874,57 +830,54 @@ BT_CONN_CB_DEFINE(conn_cb) = {
 	.disconnected = disconnected,
 };
 
-static void audio_iso_init(struct bt_audio_iso *audio_iso)
+struct audio_iso_find_params {
+	struct bt_conn *acl;
+	uint8_t cig_id;
+	uint8_t cis_id;
+};
+
+static bool audio_iso_find_func(struct bt_audio_iso *iso, void *user_data)
 {
-	/* Setup points for both sink and source
-	 * This is due to the limitation in the ISO API where pointers like
-	 * the `qos->tx` shall be initialized before the CIS is connected if
-	 * ever want to use it for TX, and ditto for RX. They cannot be
-	 * initialized after the CIS has been connected
-	 */
-	audio_iso->iso_chan.ops = &ascs_iso_ops;
-	audio_iso->iso_chan.qos = &audio_iso->iso_qos;
+	struct audio_iso_find_params *params = user_data;
+	const struct bt_audio_ep *ep;
 
-	audio_iso->iso_chan.qos->tx = &audio_iso->source_io_qos;
-	audio_iso->iso_chan.qos->tx->path = &audio_iso->source_path;
-	audio_iso->iso_chan.qos->tx->path->cc = audio_iso->source_path_cc;
+	if (iso->rx.ep != NULL) {
+		ep = iso->rx.ep;
+	} else if (iso->tx.ep != NULL) {
+		ep = iso->tx.ep;
+	} else {
+		return false;
+	}
 
-	audio_iso->iso_chan.qos->rx = &audio_iso->sink_io_qos;
-	audio_iso->iso_chan.qos->rx->path = &audio_iso->sink_path;
-	audio_iso->iso_chan.qos->rx->path->cc = audio_iso->sink_path_cc;
+	return ep->stream->conn == params->acl &&
+	       ep->cig_id == params->cig_id &&
+	       ep->cis_id == params->cis_id;
 }
 
-static struct bt_audio_iso *audio_iso_get_or_new(struct bt_ascs *ascs, uint8_t cig_id,
+static struct bt_audio_iso *audio_iso_get_or_new(struct bt_ascs *ascs,
+						 uint8_t cig_id,
 						 uint8_t cis_id)
 {
-	struct bt_audio_iso *free_audio_iso = NULL;
-
-	BT_DBG("ascs %p cig_id 0x%02x cis_id 0x%02x", ascs, cis_id, cis_id);
-
-	for (size_t i = 0; i < ARRAY_SIZE(ascs->isos); i++) {
-		struct bt_audio_iso *audio_iso = &ascs->isos[i];
-		const struct bt_audio_ep *ep;
+	struct bt_audio_iso *iso;
+	struct audio_iso_find_params params = {
+		.acl = ascs->conn,
+		.cig_id = cig_id,
+		.cis_id = cis_id,
+	};
 
-		if (audio_iso->sink_stream == NULL && audio_iso->source_stream == NULL) {
-			free_audio_iso = audio_iso;
-			continue;
-		} else if (audio_iso->sink_stream && audio_iso->sink_stream->ep) {
-			ep = audio_iso->sink_stream->ep;
-		} else if (audio_iso->source_stream && audio_iso->source_stream->ep) {
-			ep = audio_iso->source_stream->ep;
-		} else {
-			/* XXX: Stream not associated with endpoint. Should we assert??? */
-			continue;
-		}
+	iso = bt_audio_iso_find(audio_iso_find_func, &params);
+	if (iso) {
+		return iso;
```

Comment:
hmm shouldn't this return extra reference?

Reply:
You have right. I forgot to return extra reference from `bt_audio_iso_find`